src/JsonAsynchronousNodeKit/LargeJsonParser.cs

```diff
@@ -0,0 +1,281 @@
+/// Because of the use of recursion in the GetObject method, this class is also not suitable for parsing very deep json objects.
+/// </remarks>
+public class LargeJsonParser
+{
+    private const int DefaultReadBufferSize = 4096;
+    private readonly Stream stream;
+    private readonly IReadOnlyDictionary<string, PropertyHandler> handlers;
+    private readonly JsonSerializerOptions jsonSerializerOptions;
+    private byte[] buffer;
+    private JsonReaderState readerState;
+    private bool isFinalBlock;
+    private bool isParsingStarted = false;
+    private bool enumeratorActive = false;
+    private ParserStateResult? previousResultState = null;
+
+    public LargeJsonParser(
+        Stream stream,
+        IReadOnlyDictionary<string, PropertyHandler> handlers,
+        JsonSerializerOptions? jsonSerializerOptions = default,
+        int bufferSize = DefaultReadBufferSize)
+    {
+        this.stream = stream ?? throw new ArgumentNullException(nameof(stream));
+        this.handlers = handlers ?? throw new ArgumentNullException(nameof(handlers));
+        this.jsonSerializerOptions = jsonSerializerOptions ?? new JsonSerializerOptions();
+
+        this.buffer = new byte[bufferSize];
+
+        // Validate buffer is not of 0 length.
+        if (this.buffer.Length == 0)
+        {
+            throw new ArgumentException($"The {nameof(this.buffer)} value can't be null or of 0 length.");
+        }
+
+        this.stream.Position = GetStartPosition(this.stream);
+
+        if (!stream.CanRead || stream.Read(this.buffer) == 0)
+        {
+            throw new EndOfStreamException();
+        }
+
+        static int GetStartPosition(Stream stream)
+        {
+            var bom = Encoding.UTF8.Preamble.ToArray();
+            stream.Position = 0;
+            var buffer = new byte[bom.Length];
+            _ = stream.Read(buffer, 0, buffer.Length);
+
+            return Enumerable.SequenceEqual(buffer, bom) ? 3 : 0;
+        }
+    }
+
+    /// <summary>
+    /// Begin evaluating the next section of the json stream.
+    /// </summary>
+    /// <returns>A ParserStateResult object describing the section that was encountered or null if this was the final state.</returns>
+    /// <remarks>If the result object is an enumerable you MUST ensure that you've walked it entirely before calling next again.</remarks>
+    public ParserStateResult? Next()
+    {
+        if (this.enumeratorActive)
+        {
+            throw new ParserException("You must walk the entire enumerable from the previous result before calling Next() again.");
+        }
+
+        try
+        {
+            var reader = new Utf8JsonReader(this.buffer, isFinalBlock: this.isFinalBlock, this.readerState);
+
+            if (!this.isParsingStarted)
+            {
+                ParserUtils.SkipNoneTokens(this.stream, ref this.buffer, ref reader);
+
+                ParserUtils.AssertTokenType(this.stream, ref reader, JsonTokenType.StartObject);
```

Comment:
It's true that this doesn't handle a root-level-array but that's semi-intentional since for that scenario you're better off using `System.Text.Json.JsonSerializer.DeserializeAsyncEnumerable`. Since there's already a way to accomplish that and it makes the code simpler to have a grounded start to the parsing I decided not to support root-level arrays.

I'll add a comment a special exception to note this.